src/xmipp/libraries/data/monogenic.cpp

```diff
@@ -0,0 +1,712 @@
+// The thr95 represents the percentile 95 of noise distribution.
+void Monogenic::statisticsInOutBinaryMask2(const MultidimArray<double> &volS,
+		MultidimArray<int> &mask, double &meanS, double &sdS2,
+		double &meanN, double &sdN2, double &significance, double &thr95, double &NS, double &NN)
+{
+	double sumS = 0, sumS2 = 0, sumN = 0, sumN2 = 0;
+	NN = 0;
+	NS = 0;
+
+	std::vector<double> noiseValues;
+
+	FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(volS)
+	{
+		if (DIRECT_MULTIDIM_ELEM(mask, n)>=1)
+		{
+			double amplitudeValue=DIRECT_MULTIDIM_ELEM(volS, n);
+			sumS  += amplitudeValue;
+			sumS2 += amplitudeValue*amplitudeValue;
+			++NS;
+		}
+		if (DIRECT_MULTIDIM_ELEM(mask, n)==0)
+		{
+			double amplitudeValueN=DIRECT_MULTIDIM_ELEM(volS, n);
+			noiseValues.push_back(amplitudeValueN);
+			sumN  += amplitudeValueN;
+			sumN2 += amplitudeValueN*amplitudeValueN;
+			++NN;
+		}
+	}
+
+	std::sort(noiseValues.begin(),noiseValues.end());
+	thr95 = noiseValues[size_t(noiseValues.size()*significance)];
+	meanS = sumS/NS;
+	meanN = sumN/NN;
+	sdS2 = sumS2/NS - meanS*meanS;
+	sdN2 = sumN2/NN - meanN*meanN;
+}
+
+
+// SETLOCALRESOLUTIONHALFMAPS: Set the local resolution of a voxel, by
+// determining if the monogenic amplitude "amplitudeMS" is higher than
+// the threshold of noise "thresholdNoise". Thus the local resolution
+// map "plocalResolution" is set, with the resolution value "resolution"
+// or with "resolution_2". "resolution" is the resolution for with
+// the hypohtesis test is carried out, and "resolution_2" is the resolution
+// of the analisys two loops ago (see monores method)
+void Monogenic::setLocalResolutionHalfMaps(const MultidimArray<double> &amplitudeMS,
+		MultidimArray<int> &pMask, MultidimArray<double> &plocalResolutionMap,
+		double &thresholdNoise, double &resolution, double &resolution_2)
+	{
+		FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY(amplitudeMS)
+		{
+			if (DIRECT_MULTIDIM_ELEM(pMask, n)>=1)
+				if (DIRECT_MULTIDIM_ELEM(amplitudeMS, n)>thresholdNoise)
+				{
+					DIRECT_MULTIDIM_ELEM(pMask, n) = 1;
+					DIRECT_MULTIDIM_ELEM(plocalResolutionMap, n) = resolution;
+				}
+				else{
+					DIRECT_MULTIDIM_ELEM(pMask, n) += 1;
+					if (DIRECT_MULTIDIM_ELEM(pMask, n) >2)
+					{
+						DIRECT_MULTIDIM_ELEM(pMask, n) = 0;
+						DIRECT_MULTIDIM_ELEM(plocalResolutionMap, n) = resolution_2;
+					}
+				}
+		}
+	}
+
+// SETLOCALRESOLUTIONHALFMAPS: Set the local resolution of a voxel, by
+// determining if the monogenic amplitude "amplitudeMS" is higher than
+// the threshold of noise "thresholdNoise". Thus the local resolution
+// map "plocalResolution" is set, with the resolution value "resolution"
+// or with "resolution_2". "resolution" is the resolution for with
+// the hypohtesis test is carried out, and "resolution_2" is the resolution
+// of the analisys two loops ago (see monores method)
+void Monogenic::setLocalResolutionMap(const MultidimArray<double> &amplitudeMS,
```

Comment:
also duplicated code

Reply:
It is not, there is a slight difference, between this and setLocalResolutionHalfMaps